### PR TITLE
rgw: remove v4 signature special cases

### PR DESCRIPTION
**Pull request by robbat2**
V4 Signatures can be used with "Transfer-Encoding: chunked", which is
not the same as AWS V4 CHUNK encoding.

TE:Chunked is not detected as empty payload in the present case, and
worked correctly if the operation was already in the large switch
statement. Other operations not in the switch statement were wrongly
rejected as NotImplemented, when they certainly were (e.g.
DeleteObject). TE:Chunked is important because it's a trivial transform
from HTTP/2 framing, opening the door to easier HTTP/2 functionality.

If this passes s3-tests, it should be backported to all active versions.

Signed-off-by: Robin H. Johnson <rjohnson@digitalocean.com>

### CODE REVIEW

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -5633,51 +5640,7 @@ AWSGeneralAbstractor::get_auth_data_v4(const req_state* const s,
...
-        case RGW_STS_GET_SESSION_TOKEN:
-        case RGW_STS_ASSUME_ROLE:
-        case RGW_OP_PUT_BUCKET_PUBLIC_ACCESS_BLOCK:
-        case RGW_OP_GET_BUCKET_PUBLIC_ACCESS_BLOCK:
-        case RGW_OP_DELETE_BUCKET_PUBLIC_ACCESS_BLOCK:
-	case RGW_OP_GET_OBJ://s3select its post-method(payload contain the query) , the request is get-object
-          break;
-        default:
-          ldpp_dout(s, 10) << "ERROR: AWS4 completion for operation: " << s->op_type << ", NOT IMPLEMENTED" << dendl;
-          throw -ERR_NOT_IMPLEMENTED;
```
Comment (cbodley) [inline src/rgw/rgw_rest_s3.cc:5679]: > Other operations not in the switch statement were wrongly rejected as NotImplemented, when they certainly were (e.g. DeleteObject)

my interpretation of this switch statement is that it tracks the ops that call `do_aws4_auth_completion()`. we reject other ops where we haven't implemented this v4 auth completion

there are a bunch of s3 ops which we don't expect to contain a request body, and rgw replies after reading just their headers. maybe we could use a separate switch statement to track those, and perform their 'auth completion' here by comparing their `exp_payload_hash` against the empty hash `e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855`

an alternative is to add explicit calls to `do_aws4_auth_completion()` to more ops, so we can mark them as 'supported' in the existing switch

**Comment by mkll** [comment]
Any progress on it?

**Comment by LenzGr** [comment]
@robbat2 do you still intend to work on this?

**Comment by robbat2** [comment]
@LenzGr  somebody else should go fix it; afaik it's still a bug, but I don't get much time for Ceph anymore

**Comment by LenzGr** [comment]
> @LenzGr somebody else should go fix it; afaik it's still a bug, but I don't get much time for Ceph anymore

Got it, no worries. Would you mind adding an issue to the bug tracker about this?

**Comment by robbat2** [comment]
@LenzGr  filed as https://tracker.ceph.com/issues/71607

File: src/rgw/rgw_rest_s3.cc (hunk 1 above)
Comment (cbodley) [reply src/rgw/rgw_rest_s3.cc:5679]: > an alternative is to add explicit calls to `do_aws4_auth_completion()` to more ops, so we can mark them as 'supported' in the existing switch

i raised https://github.com/ceph/ceph/pull/64569 as an example that adds coverage for the s3 DeleteObject op used in the reproducers from https://tracker.ceph.com/issues/71607#note-1

we don't have a general solution for this, but can address operations on a case-by-case basis

**Pull request closed without merging**